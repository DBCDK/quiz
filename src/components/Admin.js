import React, {Component} from 'react';
import {connect} from 'react-redux';
import {
  quizDescription,
  loading,
  questionList,
  getScreen
} from '../redux/selectors';
import {
  addSection,
  deleteSection,
  editScreen,
  moveSection
} from '../redux/actions';
import DeleteIcon from '@material-ui/icons/Delete';
import DragIndicatorIcon from '@material-ui/icons/DragIndicator';
import AddIcon from '@material-ui/icons/Add';
import AppBar from '@material-ui/core/AppBar';
import Button from '@material-ui/core/Button';
import Paper from '@material-ui/core/Paper';
import Grid from '@material-ui/core/Grid';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import mustache from 'mustache';
import {DragDropContext, Droppable, Draggable} from 'react-beautiful-dnd';
import uuidv4 from 'uuid/v4';

function draggable(onMoveSection, items) {
  return (
    <DragDropContext
      onDragEnd={o =>
        onMoveSection({
          id: o.draggableId,
          from: o.source.index,
          to: o.destination.index,
          screens: items.map(o => o.id)
        })
      }
    >
      <Droppable droppableId="droppable">
        {(droppableProvided, droppableSnapshot) => (
          <div ref={droppableProvided.innerRef}>
            {items.map((item, index) => (
              <Draggable key={item.id} draggableId={item.id} index={index}>
                {(draggableProvided, draggableSnapshot) => (
                  <div
                    ref={draggableProvided.innerRef}
                    {...draggableProvided.draggableProps}
                    {...draggableProvided.dragHandleProps}
                  >
                    {item.content}
                  </div>
                )}
              </Draggable>
            ))}
            {droppableProvided.placeholder}
          </div>
        )}
      </Droppable>
    </DragDropContext>
  );
}

function quizSection({screen, doEdit, doDelete}) {
  const isDispatch = !!screen.get('dispatch', false);
  return (
    <Grid container spacing={8}>
      <Grid item xs={1}>
        <DragIndicatorIcon />
      </Grid>
      <Grid item xs={10}>
        <Button variant="contained" fullWidth={true} onClick={doEdit}>
          {screen
            .get('ui', [])
            .map(uiElem => uiElem.get('text', ''))
            .join(' ') + (isDispatch ? 'Quiz-afslutninger' : '')}
        </Button>
      </Grid>
      <Grid item xs={1}>
        {!isDispatch && (
          <Button variant="fab" aria-label="Delete" mini onClick={doDelete}>
            <DeleteIcon />
          </Button>
        )}
      </Grid>
    </Grid>
  );
}

export class Admin extends Component {
  render() {
    const questions = this.props.questions;
    return (
      <Grid container spacing={16}>
        <Grid item>[Edit description etc]</Grid>
        <Grid item>
          {!!questions.length &&
            draggable(
              this.props.moveSection,
              questions.map(q => {
                const id = q.get('_id');
                return {
                  id,
                  content: quizSection({
                    screen: q,
                    doEdit: () => this.props.editScreen(id),
                    doDelete: () => this.props.deleteSection(id)
                  })
                };
              })
            )}
        </Grid>
        <Grid item>
          <Button
            onClick={() =>
              this.props.addQuestionSection(
                questions[questions.length - 1].get('_id')
              )
            }
          >
            <AddIcon /> Spørgsmål
          </Button>
          <Button
            onClick={() => this.props.addInfoSection(questions[0].get('_id'))}
          >
            <AddIcon /> Beskrivelse, såsom intro-tekst
          </Button>
          <div />
        </Grid>
      </Grid>
    );
  }
}

function addQuestionSection(dispatch, before) {
  const questionId = uuidv4();
  const helpId = uuidv4();
  const answerId = uuidv4();
  const nextId = uuidv4();
  const screens = {
    [questionId]: {
      _id: questionId,
      nextSection: nextId,
      ui: [
        {
          type: 'image',
          image: ''
        },
        {type: 'text', text: 'Nyt spørgsmål...'},
        {
          type: 'buttonGroup',
          ui: [
            {
              type: 'button',
              text: 'Svar',
              action: {
                screen: answerId,
                increment: {
                  score: 1,
                  maxScore: 1
                }
              }
            }
          ]
        },
        {type: 'spacing'},
        {
          type: 'button',
          text: 'hjælp',
          action: {screen: helpId}
        }
      ]
    },
    [helpId]: {
      _id: helpId,
      parent: questionId,
      ui: [
        {type: 'text', text: 'Hint til spørgsmål'},
        {
          type: 'button',
          text: 'Tilbage til spørgsmålet',
          action: {screen: questionId}
        }
      ]
    },
    [answerId]: {
      _id: answerId,
      parent: questionId,
      ui: [
        {
          type: 'text',
          text: 'Feedback på besvarelsen'
        },
        {type: 'button', text: 'Fortsæt', action: {screen: nextId}}
      ],
      log: true
    }
  };
  return dispatch(addSection({before, screenId: questionId, screens}));
}

function addInfoSection(dispatch, before) {
  const sectionId = uuidv4();
  const nextId = uuidv4();
  const screens = {
    [sectionId]: {
      _id: sectionId,
      nextSection: nextId,
      ui: [
        {
          type: 'image',
          image: ''
        },
        {type: 'text', text: 'Beskrivelse, såsom intro...'},
        {type: 'spacing'},
        {
          type: 'button',
          text: 'Start',
          action: {screen: nextId}
        }
      ]
    }
  };
  return dispatch(addSection({before, screenId: sectionId, screens}));
}

export function mapStateToProps(state, ownProps) {
  return {
    questions: questionList(state).map(questionId =>
      getScreen(questionId, state).set('_id', questionId)
    ),
    loading: loading(state)
  };
}

export function mapDispatchToProps(dispatch) {
  return {
    editScreen: screen => dispatch(editScreen({screen})),
    moveSection: o => dispatch(moveSection(o)),
    deleteSection: o => dispatch(deleteSection(o)),
    addQuestionSection: before => addQuestionSection(dispatch, before),
    addInfoSection: before => addInfoSection(dispatch, before)
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Admin);
