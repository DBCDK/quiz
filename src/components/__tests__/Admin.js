import React from 'react';
import ReactDOM from 'react-dom';
import Admin from '../Admin';
import renderer from 'react-test-renderer';

it('renders without crashing', () => {
  const tree = renderer.create(<Admin />).toJSON();
  expect(tree).toMatchSnapshot();
});
