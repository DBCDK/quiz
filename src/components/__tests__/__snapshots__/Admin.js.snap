// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  className="MuiGrid-container-1 MuiGrid-spacing-xs-16-24"
>
  <div
    className="MuiGrid-item-2"
  >
    [Edit description etc]
  </div>
  <div
    className="MuiGrid-item-2"
  />
  <div
    className="MuiGrid-item-2"
  >
    <button
      className="MuiButtonBase-root-124 MuiButton-root-98 MuiButton-text-100 MuiButton-flat-103"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-99"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-127"
          color={undefined}
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
         Spørgsmål
      </span>
      <span
        className="MuiTouchRipple-root-134"
      />
    </button>
    <button
      className="MuiButtonBase-root-124 MuiButton-root-98 MuiButton-text-100 MuiButton-flat-103"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-99"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-127"
          color={undefined}
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
         Beskrivelse, såsom intro-tekst
      </span>
      <span
        className="MuiTouchRipple-root-134"
      />
    </button>
    <div />
  </div>
</div>
`;
